File: code/src/save_data_manager.py
import os
from typing import List, Tuple, Optional

import cv2
import numpy as np
import matplotlib
matplotlib.use('Agg')  # Use non-GUI backend for headless operation
import matplotlib.pyplot as plt

from logging_config import setup_logger


logger = setup_logger()


class SaveDataManager:
    """
    Manage saving operations for preloaded DICOM data (no disk reads here):
      - Save PNGs from preloaded frames
      - Write metadata JSONs from preloaded pydicom datasets
      - Create a movie from preloaded frames
    All outputs are written to the results directory.
    """

    def __init__(self, results_dir: str = "results") -> None:
        self.results_dir = results_dir
        # Check if results directory exists, create it if it doesn't
        if not os.path.exists(self.results_dir):
            os.makedirs(self.results_dir, exist_ok=True)
            logger.info(f"Created results directory: {self.results_dir}")
        else:
            logger.debug(f"Results directory already exists: {self.results_dir}")


    def save_image(self, 
                   pixel_values: np.ndarray,
                   mask: np.ndarray,
                   Value_title: str,
                   output_filename: str) -> Optional[str]:
        """
        Save pixel values as a 2D image map using a mask for spatial reference.

        Args:
            pixel_values (np.ndarray): 1D array of values for each pixel (e.g., MBF values)
            mask (np.ndarray): 2D binary mask array for spatial reference
            output_filename (str): Name of the file to save the plot

        Returns:
            Optional[str]: Path to saved file if successful, None otherwise
        """
        pixel_values = np.asarray(pixel_values)
        
        mask = np.asarray(mask)

        mask_shape = mask.shape

        # Create output directory if it doesn't exist
        output_file = os.path.join(self.results_dir, output_filename)

        # Get pixel coordinates where mask == 1
        y_coords, x_coords = np.where(mask == 1)

        # Validate that the number of pixels matches
        if len(pixel_values) != len(y_coords):
            logger.error(f"save_image: pixel_values has {len(pixel_values)} pixels, but mask has {len(y_coords)} pixels")
            return None

        # Create a 2D array for visualization (filled with NaN for pixels outside mask)
        image_2d = np.full(mask_shape, np.nan, dtype=np.float64)

        # Fill in the pixel values at their corresponding pixel locations
        for i, (y, x) in enumerate(zip(y_coords, x_coords)):
            image_2d[y, x] = pixel_values[i]

        # Create the plot
        plt.figure(figsize=(10, 8))

        # Plot the image map
        im = plt.imshow(image_2d, cmap='hot', interpolation='nearest')
        plt.title('Distribution of ' + Value_title)
        plt.xlabel('X pixel index')
        plt.ylabel('Y pixel index')

        # Add colorbar with label
        cbar = plt.colorbar(im, label=Value_title)

        # Add statistics text
        valid_values = pixel_values[pixel_values > 0]  # Exclude zeros from failed fits
        if len(valid_values) > 0:
            stats_text = f'Mean: {np.mean(valid_values):.2f}\n'
            stats_text += f'Max: {np.max(valid_values):.2f}\n'
            stats_text += f'Min: {np.min(valid_values):.2f}\n'
            plt.text(0.02, 0.98, stats_text, transform=plt.gca().transAxes,
                    verticalalignment='top', bbox=dict(boxstyle='round', facecolor='white', alpha=0.8))

        plt.tight_layout()

        # Save the plot
        try:
            plt.savefig(output_file, dpi=300, bbox_inches='tight')
            logger.info(f"Image map saved to: {output_file}")
            plt.close()
            return output_file
        except Exception as e:
            logger.error(f"save_image: error saving image: {e}")
            plt.close()
            return None


    def create_movie_from_frames(self, 
                                 frames: np.ndarray, 
                                 output_path: str = "dicom_movie.avi", 
                                 fps: int = 10, 
                                 normalize: bool = True,
                                 apply_colormap: bool = True, 
                                 codec: str = "MJPG") -> Optional[str]:
        """
        Create a movie directly from a 3D frames array (t, h, w) and save under results/.
        The movie is saved as a AVI file.
        OpenCV is used to create the movie.
        
        Args:
            frames (np.ndarray): 3D frames array (t, h, w)
            output_path (str): Output path for the movie
            fps (int): Frames per second
            normalize (bool): Normalize the frames
            apply_colormap (bool): Apply a colormap to the frames
            codec (str): Codec for the movie
        Returns:
            Optional[str]: Output path if saved successfully, else None
        """
        
        
        t_len, height, width = frames.shape
        
        output_path = os.path.join(self.results_dir, f"dicom_movie.avi")
        fourcc = cv2.VideoWriter_fourcc(*'MJPG')
        video = cv2.VideoWriter(output_path, fourcc, fps, (width, height), isColor=True)
        if not video.isOpened():
            logger.error("Failed to open video writer")
            return None
            
        # Determine normalization range
        if normalize:
            min_val = int(frames.min())
            max_val = int(frames.max())
            
        # Create the movie by traversing through the frames array in the time dimension.
        for t in range(t_len):
            arr = frames[t]
            
            if normalize:
                norm = ((arr - min_val) / (max_val - min_val) * 255).astype(np.uint8)
 
            # Available colormaps: https://docs.opencv.org/4.x/d3/d50/group__imgproc__colormap.html?utm_source=chatgpt.com
            if apply_colormap:
                frame = cv2.applyColorMap(norm, cv2.COLORMAP_TURBO)
            else:
                frame = cv2.cvtColor(norm, cv2.COLOR_GRAY2BGR)
                
            video.write(frame)
            
        video.release()
        
        logger.info(f"Movie saved: {output_path}")
        
        return output_path


    def plot_pixel_over_time(self,
                             series: np.ndarray,
                             title: str = "Pixel Timeseries",
                             y_label: str = "Signal Intensity",
                             output_filename: str = "pixel_timeseries.png") -> Optional[str]:
        """
        Convenience method: extract a pixel's associated value over time and plot it.

        Args:
            series (np.ndarray): 1D array with shape (time, value)
            title (Optional[str]): Figure title"
            y_label (str): Y-axis label
            output_filename (str): Output filename in results directory

        Returns:
            Optional[str]: Output path if saved successfully, else None
        """
        
        if series is None:
            logger.error("plot_series: series is None")
            return None
        if not isinstance(series, np.ndarray):
            try:
                series = np.asarray(series)
            except Exception as e:
                logger.error(f"plot_series: cannot convert to np.ndarray: {e}")
                return None
        if series.ndim != 1:
            logger.error("plot_series: series must be 1D")
            return None

        os.makedirs(self.results_dir, exist_ok=True)
        output_path = os.path.join(self.results_dir, output_filename)

        try:
            plt.figure(figsize=(10, 6))
            plt.plot(series, 'b-', linewidth=2)
            plt.title(title, fontsize=16)
            plt.xlabel('Time Frame', fontsize=14)
            plt.ylabel(y_label, fontsize=14)
            plt.grid(True)
            plt.tight_layout()
            plt.savefig(output_path, dpi=300, bbox_inches='tight')
            plt.close()
            logger.info(f"Series plot saved: {output_path}")
            return output_path
        except Exception as e:
            logger.error(f"plot_series: failed to save plot: {e}")
            return None


    def plot_mask(self, mask: np.ndarray, output_filename: str = "mask.png") -> Optional[str]:
        """
        Plot and save a 2D mask as a PNG file. If mask_index is provided, it is ignored for 2D masks (kept for compatibility).
        
        Args:
            mask (numpy.ndarray): Binary mask in (h, w) shape to plot.
            output_filename (str): Name of the file to save the plot.
        """
        if mask is None:
            logger.error("plot_mask: mask is None")
            return None
        if not isinstance(mask, np.ndarray):
            try:
                mask = np.asarray(mask)
            except Exception as e:
                logger.error(f"plot_mask: cannot convert to np.ndarray: {e}")
                return None

        if mask.ndim != 2:
            logger.error("plot_mask: mask must be a 2D array (h, w)")
            return None

        os.makedirs(self.results_dir, exist_ok=True)
        output_file = os.path.join(self.results_dir, output_filename)

        # Create the plot
        plt.figure(figsize=(10, 8))
        
        mask_to_plot = (1 - mask) * 255
        # Plot the mask
        im = plt.imshow(mask, cmap='gray_r', interpolation='nearest')
        
        plt.tight_layout()
        
        # Save the plot
        plt.savefig(output_file, dpi=300, bbox_inches='tight')
        logger.info(f"Mask plot saved to: {output_file}")
        plt.close()
        return output_file


if __name__ == "__main__":
    from data_loader import DataLoader
    data_loader = DataLoader()
    frames = data_loader.dicom()
    
    manager = SaveDataManager()
    manager.create_movie_from_frames(frames)
    
    # Example: plot intensity over time for the center pixel
    t_len, h, w = frames.shape
    center_row, center_col = h // 2, w // 2
    series = frames[:, center_row, center_col].astype(np.float64)
    
    manager.plot_pixel_over_time(series,
                                 title=f"Center Pixel ({center_row}, {center_col}) Intensity",
                                 y_label="Signal Intensity",
                                 output_filename="center_pixel_timeseries.png")
    
    
    mask = DataLoader(mask_file="input_data/AIF_And_Myo_Masks.tiff").mask(mask_index=0)
    manager.plot_mask(mask, output_filename="blood_pool.png")
    
    mask = DataLoader(mask_file="input_data/AIF_And_Myo_Masks.tiff").mask(mask_index=1)
    manager.plot_mask(mask, output_filename="myocardium.png")

File: codeCpp/src/SaveDataManager.cpp
#include "SaveDataManager.h"
#include "logger.h"
#include <algorithm>
#include <numeric>
#include <sstream>
#include <iomanip>

SaveDataManager::SaveDataManager(const std::string& resultsDir)
    : resultsDir(resultsDir)
{
    ensureDirectoryExists(resultsDir);
    Logger::get()->info("SaveDataManager initialized with results directory: " + resultsDir);
}

void SaveDataManager::ensureDirectoryExists(const std::string& path)
{
    if (!fs::exists(path)) {
        fs::create_directories(path);
        Logger::get()->info("Created results directory: " + path);
    } else {
        Logger::get()->debug("Results directory already exists: " + path);
    }
}

void SaveDataManager::putText(cv::Mat& img, const std::string& text, cv::Point pos,
                              double fontScale, cv::Scalar color)
{
    cv::putText(img, text, pos, cv::FONT_HERSHEY_SIMPLEX, fontScale, color, 1, cv::LINE_AA);
}

std::string SaveDataManager::saveImage(const std::vector<double>& pixelValues,
                                       const cv::Mat& mask,
                                       const std::string& valueTitle,
                                       const std::string& outputFilename)
{
    if (pixelValues.empty() || mask.empty()) {
        Logger::get()->error("saveImage: pixelValues or mask is empty");
        return "";
    }

    if (mask.type() != CV_8UC1) {
        Logger::get()->error("saveImage: mask must be CV_8UC1 (binary mask)");
        return "";
    }

    cv::Size maskShape = mask.size();
    int height = maskShape.height;
    int width = maskShape.width;

    // Get pixel coordinates where mask == 1
    std::vector<cv::Point> coords;
    for (int y = 0; y < height; ++y) {
        for (int x = 0; x < width; ++x) {
            if (mask.at<uchar>(y, x) == 1) {
                coords.push_back(cv::Point(x, y));
            }
        }
    }

    // Validate that the number of pixels matches
    if (pixelValues.size() != coords.size()) {
        Logger::get()->error("saveImage: pixelValues has " + std::to_string(pixelValues.size()) +
                             " pixels, but mask has " + std::to_string(coords.size()) + " pixels");
        return "";
    }

    // Create a 2D array for visualization (filled with NaN equivalent: -1 for invalid pixels)
    cv::Mat image2d = cv::Mat::zeros(height, width, CV_64F);
    image2d.setTo(-1); // Use -1 to indicate invalid pixels (outside mask)

    // Fill in the pixel values at their corresponding pixel locations
    for (size_t i = 0; i < coords.size(); ++i) {
        image2d.at<double>(coords[i].y, coords[i].x) = pixelValues[i];
    }

    // Find valid range (exclude -1 and zeros from failed fits)
    double minVal = std::numeric_limits<double>::max();
    double maxVal = std::numeric_limits<double>::lowest();
    double sum = 0.0;
    int validCount = 0;

    for (size_t i = 0; i < pixelValues.size(); ++i) {
        if (pixelValues[i] > 0) {
            minVal = std::min(minVal, pixelValues[i]);
            maxVal = std::max(maxVal, pixelValues[i]);
            sum += pixelValues[i];
            validCount++;
        }
    }

    // Normalize to 0-255 range for visualization
    cv::Mat normalized;
    if (maxVal > minVal) {
        image2d.copyTo(normalized);
        normalized.setTo(0, normalized < 0); // Set invalid pixels to 0
        normalized = (normalized - minVal) / (maxVal - minVal) * 255.0;
        normalized.convertTo(normalized, CV_8UC1);
    } else {
        normalized = cv::Mat::zeros(height, width, CV_8UC1);
    }

    // Apply colormap (hot colormap equivalent)
    cv::Mat colored;
    cv::applyColorMap(normalized, colored, cv::COLORMAP_HOT);

    // Draw statistics text
    std::stringstream stats;
    if (validCount > 0) {
        double mean = sum / validCount;
        stats << "Mean: " << std::fixed << std::setprecision(2) << mean << "\n";
        stats << "Max: " << std::fixed << std::setprecision(2) << maxVal << "\n";
        stats << "Min: " << std::fixed << std::setprecision(2) << minVal;
        
        // Draw text with background
        std::vector<std::string> lines;
        std::string line;
        while (std::getline(stats, line)) {
            lines.push_back(line);
        }
        
        int yPos = 30;
        for (const auto& text : lines) {
            cv::Size textSize = cv::getTextSize(text, cv::FONT_HERSHEY_SIMPLEX, 0.6, 2, nullptr);
            cv::rectangle(colored, cv::Point(10, yPos - 20), 
                         cv::Point(10 + textSize.width + 10, yPos + 5),
                         cv::Scalar(255, 255, 255), -1);
            putText(colored, text, cv::Point(15, yPos), 0.6, cv::Scalar(0, 0, 0));
            yPos += 25;
        }
    }

    // Add title
    std::string title = "Distribution of " + valueTitle;
    cv::Size titleSize = cv::getTextSize(title, cv::FONT_HERSHEY_SIMPLEX, 0.8, 2, nullptr);
    int titleX = (width - titleSize.width) / 2;
    putText(colored, title, cv::Point(titleX, 30), 0.8, cv::Scalar(255, 255, 255));

    // Save the image
    std::string outputFile = (fs::path(resultsDir) / outputFilename).string();
    try {
        cv::imwrite(outputFile, colored);
        Logger::get()->info("Image map saved to: " + outputFile);
        return outputFile;
    } catch (const cv::Exception& e) {
        Logger::get()->error("saveImage: error saving image: " + std::string(e.what()));
        return "";
    }
}

std::string SaveDataManager::createMovieFromFrames(const std::vector<cv::Mat>& frames,
                                                    const std::string& outputPath,
                                                    int fps,
                                                    bool normalize,
                                                    bool applyColormap,
                                                    const std::string& codec)
{
    if (frames.empty()) {
        Logger::get()->error("createMovieFromFrames: frames vector is empty");
        return "";
    }

    int tLen = frames.size();
    int height = frames[0].rows;
    int width = frames[0].cols;

    std::string fullOutputPath = (fs::path(resultsDir) / outputPath).string();

    // Create VideoWriter
    int fourcc = cv::VideoWriter::fourcc(codec[0], codec[1], codec[2], codec[3]);
    cv::VideoWriter video(fullOutputPath, fourcc, fps, cv::Size(width, height), true);

    if (!video.isOpened()) {
        Logger::get()->error("Failed to open video writer");
        return "";
    }

    // Determine normalization range
    double minVal = 0.0, maxVal = 255.0;
    if (normalize) {
        minVal = std::numeric_limits<double>::max();
        maxVal = std::numeric_limits<double>::lowest();
        for (const auto& frame : frames) {
            double fMin, fMax;
            cv::minMaxLoc(frame, &fMin, &fMax);
            minVal = std::min(minVal, fMin);
            maxVal = std::max(maxVal, fMax);
        }
    }

    // Create the movie by traversing through the frames
    for (int t = 0; t < tLen; ++t) {
        cv::Mat arr = frames[t].clone();

        cv::Mat norm;
        if (normalize) {
            // Handle case where all pixels have the same value
            if (maxVal == minVal) {
                norm = cv::Mat::ones(height, width, CV_8UC1) * 127; // Middle gray
            } else {
                arr.convertTo(norm, CV_8UC1, 255.0 / (maxVal - minVal), -minVal * 255.0 / (maxVal - minVal));
            }
        } else {
            arr.convertTo(norm, CV_8UC1);
        }

        cv::Mat frame;
        if (applyColormap) {
            cv::applyColorMap(norm, frame, cv::COLORMAP_TURBO);
        } else {
            cv::cvtColor(norm, frame, cv::COLOR_GRAY2BGR);
        }

        video.write(frame);
    }

    video.release();
    Logger::get()->info("Movie saved: " + fullOutputPath);
    return fullOutputPath;
}

std::string SaveDataManager::plotPixelOverTime(const std::vector<double>& series,
                                                 const std::string& title,
                                                 const std::string& yLabel,
                                                 const std::string& outputFilename)
{
    if (series.empty()) {
        Logger::get()->error("plotPixelOverTime: series is empty");
        return "";
    }

    ensureDirectoryExists(resultsDir);
    std::string outputPath = (fs::path(resultsDir) / outputFilename).string();

    try {
        int width = 1000;
        int height = 600;
        int margin = 80;
        int plotWidth = width - 2 * margin;
        int plotHeight = height - 2 * margin;

        cv::Mat plot = cv::Mat::ones(height, width, CV_8UC3) * 255; // White background

        // Find min/max for scaling
        double minVal = *std::min_element(series.begin(), series.end());
        double maxVal = *std::max_element(series.begin(), series.end());
        double range = maxVal - minVal;
        if (range == 0) range = 1.0;

        // Draw axes
        cv::line(plot, cv::Point(margin, margin), 
                 cv::Point(margin, height - margin), cv::Scalar(0, 0, 0), 2);
        cv::line(plot, cv::Point(margin, height - margin), 
                 cv::Point(width - margin, height - margin), cv::Scalar(0, 0, 0), 2);

        // Draw the line plot
        std::vector<cv::Point> points;
        for (size_t i = 0; i < series.size(); ++i) {
            int x = margin + static_cast<int>((static_cast<double>(i) / (series.size() - 1)) * plotWidth);
            int y = height - margin - static_cast<int>(((series[i] - minVal) / range) * plotHeight);
            points.push_back(cv::Point(x, y));
        }

        // Draw lines connecting points
        for (size_t i = 1; i < points.size(); ++i) {
            cv::line(plot, points[i - 1], points[i], cv::Scalar(0, 0, 255), 2); // Blue line
        }

        // Draw points
        for (const auto& pt : points) {
            cv::circle(plot, pt, 3, cv::Scalar(0, 0, 255), -1);
        }

        // Add title
        cv::Size titleSize = cv::getTextSize(title, cv::FONT_HERSHEY_SIMPLEX, 0.8, 2, nullptr);
        int titleX = (width - titleSize.width) / 2;
        putText(plot, title, cv::Point(titleX, 40), 0.8, cv::Scalar(0, 0, 0));

        // Add axis labels
        putText(plot, "Time Frame", cv::Point(width / 2 - 50, height - 20), 0.6, cv::Scalar(0, 0, 0));
        
        // Y-axis label (rotated)
        cv::Point yLabelPos(margin / 2, height / 2);
        // Note: OpenCV doesn't support rotated text easily, so we'll place it below title
        putText(plot, yLabel, cv::Point(20, height / 2), 0.6, cv::Scalar(0, 0, 0));

        // Draw grid
        for (int i = 0; i <= 10; ++i) {
            int y = margin + (plotHeight * i / 10);
            cv::line(plot, cv::Point(margin, y), cv::Point(width - margin, y), 
                     cv::Scalar(200, 200, 200), 1);
        }

        cv::imwrite(outputPath, plot);
        Logger::get()->info("Series plot saved: " + outputPath);
        return outputPath;
    } catch (const cv::Exception& e) {
        Logger::get()->error("plotPixelOverTime: failed to save plot: " + std::string(e.what()));
        return "";
    }
}

std::string SaveDataManager::plotMask(const cv::Mat& mask,
                                      const std::string& outputFilename)
{
    if (mask.empty()) {
        Logger::get()->error("plotMask: mask is empty");
        return "";
    }

    if (mask.dims != 2) {
        Logger::get()->error("plotMask: mask must be a 2D array");
        return "";
    }

    ensureDirectoryExists(resultsDir);
    std::string outputFile = (fs::path(resultsDir) / outputFilename).string();

    try {
        // Convert mask to displayable format (invert for grayscale display)
        cv::Mat maskToPlot;
        if (mask.type() == CV_8UC1) {
            maskToPlot = (1 - mask) * 255; // Invert: 0 becomes 255, 1 becomes 0
        } else {
            mask.convertTo(maskToPlot, CV_8UC1);
        }

        // Convert to BGR for saving (grayscale visualization)
        cv::Mat colored;
        cv::cvtColor(maskToPlot, colored, cv::COLOR_GRAY2BGR);

        cv::imwrite(outputFile, colored);
        Logger::get()->info("Mask plot saved to: " + outputFile);
        return outputFile;
    } catch (const cv::Exception& e) {
        Logger::get()->error("plotMask: failed to save plot: " + std::string(e.what()));
        return "";
    }
}

